middleware/session/middleware.go

```diff
@@ -0,0 +1,188 @@
+package session
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/gofiber/fiber/v3"
+	"github.com/gofiber/fiber/v3/log"
+)
+
+// key for looking up session middleware in request context
+const key = 0
+
+// Session defines the session middleware configuration
+type MiddlewareConfig struct {
+	// Next defines a function to skip this middleware when returned true.
+	//
+	// Optional. Default: nil
+	Next func(c fiber.Ctx) bool
+
+	// Store defines the session store
+	//
+	// Required.
+	Store *Store
+
+	// ErrorHandler defines a function which is executed for errors
+	//
+	// Optional. Default: nil
+	ErrorHandler func(*fiber.Ctx, error)
+}
+
+type Middleware struct {
+	config     MiddlewareConfig
+	Session    *Session
+	ctx        *fiber.Ctx
+	hasChanged bool
+	mu         sync.RWMutex
+}
+
+// Middleware pool
+var middlewarePool = &sync.Pool{
+	New: func() any {
+		return &Middleware{}
+	},
+}
+
+// Session is a middleware to manage session state
+//
+// Session middleware manages common session state between requests.
+// This middleware is dependent on the session store, which is responsible for
+// storing the session data.
+func NewMiddleware(config MiddlewareConfig) fiber.Handler {
```

Comment:
I think it's better if we use New() here. Maybe we can rename current New to NewStore, NewStorage etc.

Reply:
I believe that uses only e `New` or `NewSession`

Reply:
How should we handle the interaction between Session and Middleware? Here's the potential issue: if the middleware is in use and the session for the same user is retrieved from the store, any changes made and saved in the handler would be overwritten by the middleware.

We could add a check in the store. If the middleware is in the context, it could pass that session of `Store.Get(c)` and `Session.Save()`, and it could `noop` if it's the middleware session.

Reply:
@sixcolors it makes sense to me. I believe that follow with your point.